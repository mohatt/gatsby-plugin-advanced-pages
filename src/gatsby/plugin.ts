import type { GatsbyNode, PluginOptions as GatsbyPluginOptions } from 'gatsby'
import { options, reporter } from './util'
import { getOptionsSchema } from './schema'
import type { PluginOptions } from './index'

// Validates user-defined options against schema
// runs before onPreBootstrap
export const pluginOptionsSchema: GatsbyNode['pluginOptionsSchema'] = ({ Joi }) => {
  return getOptionsSchema(Joi)
}

// Initializes plugin state
export const onPluginInit: GatsbyNode['onPluginInit'] = (
  args,
  pluginOptions: GatsbyPluginOptions & PluginOptions,
) => {
  // Default values for options
  const defaultOptions: PluginOptions = {
    basePath: '/',
    pages: [],
    template: null,
    directories: {
      templates: './src/templates',
      helpers: './gatsby/pages',
    },
    pagination: {
      limit: 10,
      suffix: '/page/:page',
    },
    typeNames: {
      page: 'Page',
      pageRoute: 'PageRoute',
    },
  }

  // Initializes and validates options
  // Only runs once at bootstrap
  options.initialize({
    store: args.store,
    defaultOptions,
    pluginOptions,
  })

  // Initializes plugin reporter
  reporter.initialize(args.reporter)
}

// Creates a webpack alias for the plugin cache directory
// Usually located at <root>/.cache/caches/<plugin>
export const onCreateWebpackConfig: GatsbyNode['onCreateWebpackConfig'] = ({ actions, cache }) => {
  actions.setWebpackConfig({
    resolve: {
      alias: {
        'gatsby-plugin-advanced-pages-cache': cache.directory,
      },
    },
  })
}

export const createSchemaCustomization: GatsbyNode['createSchemaCustomization'] = ({
  actions,
  schema,
}) => {
  const { createTypes } = actions
  const { page, pageRoute } = options.get('typeNames')

  createTypes([
    schema.buildObjectType({
      name: pageRoute,
      fields: {
        name: 'String!',
        path: 'String!',
      },
    }),
    schema.buildObjectType({
      name: page,
      fields: {
        title: 'String!',
        templateName: 'String',
        template: 'String',
        helper: 'String',
        data: 'JSON',
        routes: `[${pageRoute}!]!`,
      },
      interfaces: ['Node'],
    }),
  ])
}

export { default as sourceNodes } from './source-nodes'
export { default as createPages } from './create-pages'
