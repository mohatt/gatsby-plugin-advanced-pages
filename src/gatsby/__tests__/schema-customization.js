import { createSchemaCustomization } from '../plugin'
import { mountOptions } from '../../../test/node-utils'

describe('createSchemaCustomization', () => {
  const createTypes = jest.fn()
  const actions = { createTypes }
  const schema = { buildObjectType: (obj) => obj }

  it('correctly creates schema definitions', () => {
    mountOptions()

    let error = null
    try {
      createSchemaCustomization({
        actions,
        schema
      })
    } catch (e) {
      error = e
    }

    expect(error).toBeNull()
    expect(createTypes.mock.calls).toMatchSnapshot()
  })
})
