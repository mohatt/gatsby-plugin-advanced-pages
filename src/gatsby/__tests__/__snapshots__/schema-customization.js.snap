// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSchemaCustomization correctly creates schema definitions 1`] = `
[
  [
    [
      {
        "fields": {
          "name": "String!",
          "path": "String!",
        },
        "name": "PageRoute",
      },
      {
        "fields": {
          "data": "JSON",
          "helper": "String",
          "routes": "[PageRoute!]!",
          "template": "String",
          "templateName": "String",
          "title": "String!",
        },
        "interfaces": [
          "Node",
        ],
        "name": "Page",
      },
    ],
  ],
]
`;
